### ISSUE TITLE
Support for Hashmaps in EMF Model

### ISSUE THREAD

**Issue by buchgeher** [atlanmod/NeoEMF#4]
Hi,

in my model I am making use of maps as described in https://wiki.eclipse.org/EMF/FAQ#How_do_I_create_a_Map_in_EMF.3F 

When I try to access a map I am getting a class cast exception. I'm not really sure if this problem is related to NeoEMF, but the exception does not occur when generating/using a standard EMF model.

org.osgi.framework.BundleException: Exception in neoemfsample.Activator.start() of bundle neoemfsample.
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:792)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:721)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:936)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:319)
	at org.eclipse.osgi.container.Module.doStart(Module.java:571)
	at org.eclipse.osgi.container.Module.start(Module.java:439)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1562)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
Caused by: java.lang.ClassCastException: org.eclipse.emf.ecore.impl.EStoreEObjectImpl$BasicEStoreEList cannot be cast to org.eclipse.emf.common.util.EMap
	at sample.impl.SampleModelImpl.getMap(SampleModelImpl.java:82)
	at neoemfsample.Activator.test(Activator.java:72)
	at neoemfsample.Activator$1.run(Activator.java:61)
	at neoemfsample.Activator.start(Activator.java:63)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:771)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:764)
	... 12 more
Root exception:
java.lang.ClassCastException: org.eclipse.emf.ecore.impl.EStoreEObjectImpl$BasicEStoreEList cannot be cast to org.eclipse.emf.common.util.EMap
	at sample.impl.SampleModelImpl.getMap(SampleModelImpl.java:82)
	at neoemfsample.Activator.test(Activator.java:72)
	at neoemfsample.Activator$1.run(Activator.java:61)
	at neoemfsample.Activator.start(Activator.java:63)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:771)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:764)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:721)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:936)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:319)
	at org.eclipse.osgi.container.Module.doStart(Module.java:571)
	at org.eclipse.osgi.container.Module.start(Module.java:439)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1562)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)

**Comment by gdaniel** [comment]
Hi,
Can you provide an archive with the code that throws this error?

**Comment by buchgeher** [comment]
My sample plug-in can be found here:
https://dl.dropboxusercontent.com/u/366920/neoemfmapsample.zip

**Comment by gdaniel** [comment]
It is effectively a problem in our implementation: we do not consider the use of EMaps.
In the current state all the multi-valued features are handled with Lists, that's why you get this ClassCastException.
I am having a look on it but it may take time to provide and test an implementation for the two backends.

Gwendal

**Comment by buchgeher** [comment]
Hi Gwendal,

thanks for explaining the issue. I'm currently in the process of evaluating if NeoEMF can be used as persistence backend in our project. We have to store a large EMF model that will likely grow significantly in the near future. NeoEMF is a very interesting and approach, especially because similar approaches already turned out as not appropriate. In our our model we use EMaps multiple times. Since the model is already used in production we cannot change these EMaps.
I know this is probably difficult but can you make a rough estimation when you can provide an implementation that supports EMaps? Is this a matter of weeks or months? I would need this information for our internal project planning. 

Best Regards
Georg

**Comment by gdaniel** [comment]
Hi Georg,

I understand your request, and I try to see how we can provide a solution as quick as possible.
Please note that I maintain NeoEMF on my free time, and it may take some time to provide a working solution, according to my availabilities. 

I keep you in touch with that.

Gwendal

**Comment by gdaniel** [comment]
Hi again,

I have pushed a new version of the update site that includes EMap support.
We support them the same way they are implemented in [MinimalEStoreEObjectImpl](https://github.com/eclipse/emf/blob/master/plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/impl/MinimalEStoreEObjectImpl.java).

I also added test-commons and tests folders in the repository. For the moment it contains a model and a test case for your issue. I hope this tests could give you some usage examples (in addition to the wiki which is under construction).

Best regards,

Gwendal
